Bug Fix: Update hook_views_data_alter() to fix issue #789

guignonv opened the pull request:
Bug Fix

Issue #789 

## Description
Allows to add a relationship between Chado stock and associated Drupal entities on views. Then, it's possible to have a view on Chado stock that displays both data from Chado tables and entity field data from Drupal.

## Testing?
Create a view on "Chado Stock", then add relationship: new Stock relationships now exist between Stock and Entity allowing to display Drupal fields associated to the Chado stock.

## Additional Notes (if any):
I think there are still some adjustments to be made: some $data[some_table]['table']['join'][...] may not be needed and I did not succeed to add the relationship from a Tripal entity  to a Chado object. However, from Chado to Tripal entity, it is working. At least, it's better than current code that does nothing!
Also tests should be added.

guignonv committed:
Update hook_views_data_alter() to fix issue #789

I think there are still some adjustments to be made: some $data[<table>]['table']['join'] may not be needed and I did not succeed to add the relationship from a Tripal entity  to a Chado object. However, from Chado to Tripal entity, it is working. At least, it's better than current code that does nothing!

shawnawsu commented:
Sorry for the delay in reviewing this I was out with the flu. The code looks fine and I don't get any errors or problems when on the branch. I can't verify that the view works because I don't actually have any data in the stock tables in my chado db. @bradfordcondon if you've got a moment to verify the view actually works I think that would be ideal before pushing the code out.

bradfordcondon commented:
Hi @guignonv ,

I created two Cultivar (germplasm Variety) entities, and linked them via a relationship.  

When creating a view of Cultivar (Germplasm Variety), the relationships show up as available fields on your branch, but also on `7.x-3.x`.  However on both branches its just a jumbled mess in the view itself.  Is there some specific view configuration i need to see the problem?  Would you mind exporting it so i can test it more easily?

You also keep talking about Chado Stocks. Do you mean the Tripal 2 node type?  

I dont understand what doesn't work and waht this PR is fixing.

guignonv commented:
@bradfordcondon: Now, to understand it better, create a new Drupal site and enable modules views, tripal, tripal_chado and tripal_chado_views.
Initialize an empty Chado with Tripal.
Now add:
```
% INSERT INTO chado.organism (abbreviation, genus, species, common_name, infraspecific_name, type_id, comment)
  VALUES ('M.acuminata', 'Musa', 'acuminata', 'banana', '', NULL, '');
% INSERT INTO chado.db (name, description, urlprefix, url)
  VALUES ('MusaBase', 'MusaBase', 'https://musabase.org/stock/', 'https://musabase.org/');
% INSERT INTO chado.db (name, description, urlprefix, url)
  VALUES ('BrapiMusaBase', 'MusaBase BrAPI', 'https://musabase.org/brapi/v1/germplasm-search?accessionNumber=', 'https://musabase.org/');
% INSERT INTO chado.db (name, description, urlprefix, url)
  VALUES ('BGH', 'Banana Genome Hub', 'https://banana-genome-hub.southgreen.fr/organism/Musa/', 'https://banana-genome-hub.southgreen.fr/');
% INSERT INTO chado.dbxref(db_id, accession, version, description)
  SELECT d.db_id, '65868/view', '', '' FROM chado.db d WHERE d.name = 'MusaBase';
% INSERT INTO chado.dbxref(db_id, accession, version, description)
  SELECT d.db_id, 'ITC0609', '', '' FROM chado.db d WHERE d.name = 'BrapiMusaBase';
% INSERT INTO chado.dbxref(db_id, accession, version, description)
  SELECT d.db_id, 'acuminata', '', '' FROM chado.db d WHERE d.name = 'BGH';
% INSERT INTO chado.stock (dbxref_id, organism_id, name, uniquename, description, type_id, is_obsolete)
  SELECT 1, o.organism_id, 'Pahang', 'ACC001', '', ct.cvterm_id, FALSE
  FROM chado.cvterm ct JOIN chado.cv c ON c.cv_id = ct.cv_id AND c.name = 'germplasm_ontology',
    chado.organism o
  WHERE
    o.species = 'acuminata'
    AND ct.name = 'accession'
  ;
% INSERT INTO stock_dbxref(stock_id, dbxref_id, is_current)
  SELECT s.stock_id, dx.dbxref_id, TRUE
  FROM chado.stock s, chado.dbxref dx JOIN chado.db d USING (db_id)
  WHERE d.name = 'MusaBase'
  AND dx.accession = 'ITC0609'
  AND s.uniquename = 'ACC001';
% INSERT INTO stock_dbxref(stock_id, dbxref_id, is_current)
  SELECT s.stock_id, dx.dbxref_id, TRUE
  FROM chado.stock s, chado.dbxref dx JOIN chado.db d USING (db_id)
  WHERE d.name = 'BGH'
  AND dx.accession = 'acuminata'
  AND s.uniquename = 'ACC001';
```
and publish "Organism" and "Germplasm Accession".

Create a new view on "Germplasm Accession" as a page that displays a table.
Now the goal to acchieve is to display something like that:
```
 Name | Link to Musabase | Link to BGH | BrAPI data at MusaBASE
------+------------------+-------------+------------------------
[Pahang](/bio_data/2)|        [65868/view](https://musabase.org/stock/65868/view)|    [acuminata](https://banana-genome-hub.southgreen.fr/organism/Musa/acuminata)| [https://musabase.org/brapi/v1/germplasm-search?accessionNumber=ITC0609](https://musabase.org/brapi/v1/germplasm-search?accessionNumber=ITC0609)
```
Where "Pahang" links to the Drupal content (something like "/bio_data/2"),
"65868/view" links to "https://musabase.org/stock/65868/view" and
"acuminata" links to "https://banana-genome-hub.southgreen.fr/organism/Musa/acuminata"
As far as I know, it is not possible as you can't have several "Germplasm Accession: Cross Reference Database URL (Cross Reference Database URL)" fields, each of them being filtered on a given cross ref database.
The way to achieve this would be to add cross references using the "relationships" feature of views and then use relationship-specific filter criteria but you can't add any kind of views relationship on this type of content. With Tripal 2, it was possible to add a relationship to the associated Chado Stock and then add the other (Chado-side) relationships.

Now, create a new view on "Chado stock". You can add 3
"Chado Stock: stock_id => Stock Dbxref (stock_id)" relationships and then 3
"Chado Stock Dbxref: dbxref_id => Dbxref (dbxref_id)" each of them associated to
one of the previous "Chado Stock: stock_id => Stock Dbxref (stock_id)" relationships.
Then, you can filter each of those 3 "Chado Dbxref" on their "Chado Dbxref: Db Id".
One for "MusaBase", one for "BrapiMusaBase" and one for "BGH".
Then you can display the dbxref-related and db-related fields you want according
to the view relationship you want.
With the patch, you can now also link the stock (Germplasm Accession) to the
Drupal content using the views relationship "Stock: Stock => Entity" which
could not be achieved before (but was available with Tripal 2).

But maybe you never used views relationships before and that's why you did not
understand what I was trying to do?

Note: the use of dbxref relationship here is just one use case but there are many others with stock-organism-phylonode_organism-phylonode, stock-stock_cvterm, stock-stock_relationship-stock and stock-nd_experiment_stock-nd_* tables...

bradfordcondon commented:
>But maybe you never used views relationships before and that's why you did not understand what I was trying to do?

This is true.  Would be awesome if someone else could step in and do the functional review for this reason, but I'll do my best.
  Thanks for the description I'll try this again but will take me a few days to get to it unfortunately.

laceysanderson commented on tripal_chado_views/tripal_chado_views.views.inc:
```diff
@@ -600,84 +600,97 @@ function tripal_chado_views_views_data_tripal_views_tables($data) {
+    // Ids we'll use for queries.
+    $base_table = $bundle->data_table;
+    $linker_table = 'chado_bio_data_' . $bundle->bundle_id;
+    $base_title = ucwords(str_replace('_', ' ', $base_table));
```
This needs to be the bundle name. Currently I end up with 4 different `Stock: Stock => Entity` relationships and I need to use trial and error to determine which one is right for my content. This is also only a good solution if all my content is of one Tripal Content Type. Otherwise, it becomes difficult since you need to use a conditional to set a link on the stock name.

This is what I currently see in my relationship list on a fresh Tripal install:
![screen shot 2019-03-06 at 7 12 13 pm](https://user-images.githubusercontent.com/1566301/53925206-b9c28e80-4044-11e9-9e6e-26fcf8408f12.png)

I was able to confirm that one I selected the right `Stock: Stock => Entity` for my bundle that I got the entity links 🎉

guignonv replied:
It's a bit weird because I had only "Stock: Stock => Entity" on my system but I re-installed everything and restarted from scratch and that's right, 4 relationships appeared! :-/ That's not what's wanted. I'll try see what's going on.

guignonv committed:
Fixed relationship labels

Fixed relationship labels and did some code refactoring. Added relationships to Tripal bundles but the TripalFiledQuery object does not seem to fetch the joined values.

laceysanderson commented on tripal_chado_views/tripal_chado_views.views.inc:
```diff
@@ -596,88 +596,130 @@ function tripal_chado_views_views_data_tripal_views_tables($data) {
+          'group' => $base_table_label,
+          'title' => $base_table_label,
+          'help' => t(
+            "Links @base_table_label to its @bundle_label.",
```
I think we should mention Tripal Content Type (Entity) to make it easier to search for all Tripal Content type relationships.
```suggestion
            "Links @base_table_label to its Tripal Content Type '@bundle_label'.",
```

laceysanderson reviewed (approved):
I've functionally tested this PR and it allows a Chado Stock View to connect to it's Tripal Content Type. It is limited to a single content type per relationships so we likely still want a specialized handler at some point to allow linking to all content types with a given base at once.

Here is the updated view of the relationships:
![screen shot 2019-03-07 at 8 18 08 am](https://user-images.githubusercontent.com/1566301/53963309-a77f3980-40b2-11e9-9e7e-73ebe83dd199.png)

laceysanderson commented:
@shawnawsu can you reconfirm by using the default Tripal content types for stock (e.g. generated germplasm) to create content in your test site? Here is a view export for testing:
```
$view = new view();
$view->name = 'stock_test';
$view->description = '';
$view->tag = 'default';
$view->base_table = 'stock';
$view->human_name = 'stock test';
$view->core = 7;
$view->api_version = '3.0';
$view->disabled = FALSE; /* Edit this to true to make a default view disabled initially */

/* Display: Master */
$handler = $view->new_display('default', 'Master', 'default');
$handler->display->display_options['title'] = 'stock test';
$handler->display->display_options['use_more_always'] = FALSE;
$handler->display->display_options['access']['type'] = 'none';
$handler->display->display_options['cache']['type'] = 'none';
$handler->display->display_options['query']['type'] = 'views_query';
$handler->display->display_options['exposed_form']['type'] = 'basic';
$handler->display->display_options['pager']['type'] = 'some';
$handler->display->display_options['pager']['options']['items_per_page'] = '10';
$handler->display->display_options['style_plugin'] = 'table';
/* Relationship: Chado Stock: Chado Stock => Generated Germplasm (breeding Cross) */
$handler->display->display_options['relationships']['stock_id']['id'] = 'stock_id';
$handler->display->display_options['relationships']['stock_id']['table'] = 'chado_bio_data_22';
$handler->display->display_options['relationships']['stock_id']['field'] = 'stock_id';
/* Field: Chado Stock: Stock Id */
$handler->display->display_options['fields']['stock_id']['id'] = 'stock_id';
$handler->display->display_options['fields']['stock_id']['table'] = 'stock';
$handler->display->display_options['fields']['stock_id']['field'] = 'stock_id';
/* Field: Chado Stock: Name */
$handler->display->display_options['fields']['name']['id'] = 'name';
$handler->display->display_options['fields']['name']['table'] = 'stock';
$handler->display->display_options['fields']['name']['field'] = 'name';
/* Field: Chado Stock: Uniquename */
$handler->display->display_options['fields']['uniquename']['id'] = 'uniquename';
$handler->display->display_options['fields']['uniquename']['table'] = 'stock';
$handler->display->display_options['fields']['uniquename']['field'] = 'uniquename';
/* Field: Tripal Content: URL */
$handler->display->display_options['fields']['url']['id'] = 'url';
$handler->display->display_options['fields']['url']['table'] = 'views_entity_TripalEntity';
$handler->display->display_options['fields']['url']['field'] = 'url';
$handler->display->display_options['fields']['url']['relationship'] = 'stock_id';
$handler->display->display_options['fields']['url']['link_to_entity'] = 0;

/* Display: Page */
$handler = $view->new_display('page', 'Page', 'page');
$handler->display->display_options['path'] = 'stock-test';
```

laceysanderson committed:
Update tripal_chado_views/tripal_chado_views.views.inc

Co-Authored-By: guignonv <guignonv@users.noreply.github.com>

guignonv committed:
Merge branch '7.x-3.x' into patch-1

shawnawsu commented:
I can confirm that I get a view that returns content which I believe to be the correct content, though I've included a screenshot to confirm this is correct?
<img width="874" alt="Screen Shot 2019-03-22 at 1 24 50 PM" src="https://user-images.githubusercontent.com/24374002/54841671-304ac780-4ca6-11e9-8a02-f5eeb0c173ad.png">

@laceysanderson If you confirm this is correct I can add my approval tag to the PR.

laceysanderson commented:
Thanks @guignonv for the fix!
